flowpipe/graph.py

```diff
@@ -275,30 +275,49 @@ def _evaluate_threaded(self, nodes_to_evaluate, max_workers=None):
         """Evaluate each node in a new thread."""
         log.debug("{0} evaluating {1} nodes in threading mode.".format(
             self.name, len(nodes_to_evaluate)))
+
         def node_runner(node):
             """Run a node's evaluate method and return the node."""
             node.evaluate()
             return node
 
-        running_futures = []
+        running_futures = {}
         with futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
             while nodes_to_evaluate or running_futures:
+                log.debug("Iterating thread submission with {0} nodes to "
+                          "evaluate and {1} running futures".format(
+                              len(nodes_to_evaluate), len(running_futures)))
                 # Submit new nodes that are ready to be evaluated
                 not_submitted = []
                 for node in nodes_to_evaluate:
                     if not any(n.is_dirty for n in node.upstream_nodes):
                         fut = executor.submit(node_runner, node)
-                        running_futures.append(fut)
+                        running_futures[node.name] = fut
                     else:
                         not_submitted.append(node)
                 nodes_to_evaluate = not_submitted
 
+                # A deadlock situation:
+                # No nodes running means no nodes can turn clean but nodes on
+                # nodes_to_evaluate not submitted means dirty upstream nodes
+                # and while loop will never terminate
+                if nodes_to_evaluate and not running_futures:
+                    for node in nodes_to_evaluate:
+                        dirty_upstream = [nn.name for nn in node.upstream_nodes
+                                          if nn.is_dirty]
+                        log.debug("Node to evaluate: {0} ".format(node.name) +
+                                  "- Dirty upstream nodes:\n" +
+                                  "\n".join(dirty_upstream))
```

Comment:
This assembles a multiline log that will inform you of which node is stuck and which upstream nodes are at fault. e.g.:

```
Node to evaluate: Finalizer - Dirty upstream nodes:
Upstream Node 1
Upstream Node 2

Node to evaluate: Node 2 - Dirty upstream nodes:
Upstream Node 1
```